telemetry/Tracks/class-tracks-event.php

```diff
@@ -0,0 +1,275 @@
+<?php
+/**
+ * Telemetry: Tracks Event class
+ *
+ * @package Automattic\VIP\Telemetry
+ */
+
+declare(strict_types=1);
+
+namespace Automattic\VIP\Telemetry;
+
+use stdClass;
+use WP_Error;
+
+/**
+ * Class that creates and validates Tracks events.
+ *
+ * @since 3.12.0
+ */
+class Tracks_Event {
+	/**
+	 * Event name prefix.
+	 */
+	protected const EVENT_NAME_PREFIX = 'viptelemetry_';
+
+	/**
+	 * Event name regex.
+	 */
+	protected const EVENT_NAME_REGEX = '/^(([a-z0-9]+)_){1}([a-z0-9_]+)$/';
+
+	/**
+	 * Property name regex.
+	 */
+	protected const PROPERTY_NAME_REGEX = '/^[a-z_][a-z0-9_]*$/';
+
+	/**
+	 * Variable containing the event's data or an error if one was encountered
+	 * during the event's creation.
+	 *
+	 * @var stdClass|WP_Error
+	 */
+	protected $data = null;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string                            $event_name The event's name.
+	 * @param array<string, mixed>|array<empty> $event_properties Any properties included in the event.
+	 */
+	public function __construct( string $event_name, array $event_properties ) {
+		$event_data        = static::process_properties( $event_name, $event_properties );
+		$validation_result = static::get_event_validation_result( $event_data );
+
+		$this->data = $validation_result ?? $event_data;
+	}
+
+	/**
+	 * Returns the event's data.
+	 *
+	 * @return stdClass|WP_Error Event object if the event was created successfully, WP_Error otherwise.
+	 */
+	public function get_data() {
+		return $this->data;
+	}
+
+	/**
+	 * Returns whether the event can be recorded.
+	 *
+	 * @return bool|WP_Error True if the event is recordable.
+	 *                       False if the event is not recordable.
+	 *                       WP_Error if the event is generating an error.
+	 */
+	public function is_recordable() {
+		if ( ! Telemetry_System::is_wpadmin_telemetry_allowed() ) {
+			return false;
+		}
+
+		// Don't record events during unit tests and CI runs.
+		if ( 'wptests_capabilities' === wp_get_current_user()->cap_key ) {
+			return false;
+		}
```

Comment:
This feels odd...

Reply:
This is to stop events from being recorded during tests, is this not the correct way to do it? Or is it just unnecessary?

Reply:
More just that it's happening in an unusual spot in the code. Feels like we should do this as late as possible (i.e. before the request is sent).

Reply:
The `is_recordable` method is being called by methods like `record_event_synchronously` which are the methods that makes the request. This seems pretty late to me though 🤔

Reply:
[This](https://github.com/Automattic/vip-block-data-api/blob/trunk/src/analytics/analytics.php#L99) is one way that we limited such tracking from happening (for MC Stats) to only production sites. This would be worth considering so it's done pretty early on.